projects/optic/src/commands/diff/diff-all.ts

```diff
@@ -0,0 +1,452 @@
+import { Command } from 'commander';
+import { OpticCliConfig, VCS } from '../../config';
+import { wrapActionHandlerWithSentry } from '@useoptic/openapi-utilities/build/utilities/sentry';
+import {
+  findOpenApiSpecsCandidates,
+  resolveGitRef,
+} from '../../utils/git-utils';
+import { getFileFromFsOrGit, ParseResult } from '../../utils/spec-loaders';
+import { logger } from '../../logger';
+import { OPTIC_URL_KEY } from '../../constants';
+import { compute } from './compute';
+import chalk from 'chalk';
+import {
+  flushEvents,
+  trackEvent,
+} from '@useoptic/openapi-utilities/build/utilities/segment';
+import { getAnonId } from '../../utils/anonymous-id';
+import open from 'open';
+import { compressData } from './compressResults';
+import {
+  generateChangelogData,
+  generateComparisonLogs,
+  jsonChangelog,
+  logComparison,
+  terminalChangelog,
+} from '@useoptic/openapi-utilities';
+
+const usage = () => `
+  optic diff-all
+  optic diff-all --compare-from main --compare-to feat/new-api --check --web --ruleset @org/example-ruleset`;
+
+const helpText = `
+Example usage:
+  Diff all specs with \`x-optic-url\` in the current repo against HEAD~1
+  $ optic diff-all
+
+  Diff all specs with \`x-optic-url\` in the current repo from main to feature/1
+  $ optic diff-all --compare-from main --compare-to feature/1
+
+  Diff all specs with a ruleset, run checks and open up in a web browser
+  $ optic diff-all --ruleset @org/example-ruleset --web --check
+  `;
+
+export const registerDiffAll = (cli: Command, config: OpticCliConfig) => {
+  cli
+    .command('diff-all')
+    .configureHelp({
+      commandUsage: usage,
+    })
+    .addHelpText('after', helpText)
+    .description('Run a diff on all specs with `x-optic-url`')
+    .option(
+      '--compare-to <compare-to>',
+      'the head ref to compare against. Defaults to the current working directory'
+    )
+    .option(
+      '--compare-from <compare-from>',
+      'the base ref to compare against. Defaults to HEAD~1',
+      'HEAD~1'
+    )
+    .option(
+      '--ruleset <ruleset>',
+      'run comparison with a locally defined ruleset, if not set, looks for the ruleset on the [x-optic-ruleset] key on the spec, and then the optic.dev.yml file.'
+    )
+    .option('--check', 'enable checks', false)
+    .option('--web', 'view the diff in the optic changelog web view', false)
+    .option('--json', 'output as json', false)
+    .action(wrapActionHandlerWithSentry(getDiffAllAction(config)));
+};
+
+type DiffAllActionOptions = {
+  compareTo?: string;
+  compareFrom: string;
+  ruleset?: string;
+  check: boolean;
+  web: boolean;
+  json: boolean;
+};
+
+// Match up the to and from candidates
+// This will return the comparisons we can try to run
+function matchCandidates(
```

Comment:
Here we pull both the sets of `from sha` and `to sha` and match up the results.

We later disregard certain files once we read the contents (below)